File: Binary Search Tree/binarysearchtree.js
class Node {
    constructor(val) {
        this.val = val;
        this.right = null;
        this.left = null;
    }
}

class BinarySearchTree{
    constructor() {
        this.root = null;
    }

    traverse(root) {
        if (root == null) {
            return
        }
        this.traverse(root.left);
        console.log(root.val);
        this.traverse(root.right);
    }

    search(val) {
        if (this.root == null) {
            console.log("The tree has no values");
            return
        }
        var node = this.root
        while (node != null) {
            if (val < node) {
                node = node.left;
                continue
            }
            else if (val > node) {
                node = node.right
                continue
            }
            else {
                return node
            }
        }
        console.log("Value was not found in tree")
        return
    }

    insert(val) {
        var new_node = new Node(val)
        if (this.root == null) {
            this.root = new_node
            return
        }

        var parent = this.root

        while (true) {
            if (val < parent.val) {
                if (parent.left == null) {
                    parent.left = new_node;
                    return
                }
                else {
                    parent = parent.left
                }
            }
            else if (val > parent.val) {
                if (parent.right == null) {
                    parent.right = new_node;
                    return
                }
                else {
                    parent = parent.right
                }
            }
            else {
                console.log("Function does not support duplicates");
                return
            }
        }
    }

    delete(val) {
        if (this.root == null) {
            console.log("Tree does not contain value");
            return
        }

        var parent = self.root;
        var node = self.root;

        while (node != null) {
            if (val < node.val) {
                parent = node;
                node = node.left;
            }
            else if (val > node.val) {
                parent = node
                node = node.right
            }
            else {
                if (node.left == null && node.right == null) {
                    if (parent.left == node) {
                        parent.left = null
                        return
                    }
                    else {
                        parent.right = null
                        return
                    }
                }
                else if (parent.left == node) {
                    if (node.left == null) {
                        parent.left = node.right
                    }
                    else if (node.right == null) {
                        parent.left = node.left
                    }
                    else {
                        parent.left = node.right
                    }
                }
                else if (parent.right == node) {
                    if (node.left == null) {
                        parent.right = node.right
                    }
                    else if (node.right == null) {
                        parent.right = node.left
                    }
                    else {
                        parent.right = node.left
                    }
                }
            }
        }

    }
}

File: Binary Search Tree/binarysearchtree.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None

    def search(self, val):
        if self.root == None:
            return -1

        def iterate(root, val):
            if root is None:
                return -1

            if root.val == val:
                return root
            
            if root.val > val:
                return iterate(root.left, val)
            if root.val < val:
                return iterate(root.right, val)

        return iterate(self.root, val)

    def inorder(self, root):
        if root is None:
            return
        
        self.inorder(root.left)
        print(root.val)
        self.inorder(root.right)

    # Struggles when the values are duplicated
    def insert(self, val):
        new_node = Node(val)
        if self.root is None:
            self.root = new_node
            return

        parent = self.root

        while True:
            if val < parent.val:  # Traverse to the left subtree
                if parent.left is None:  # Insert if left child is empty
                    parent.left = new_node
                    return
                parent = parent.left  # Move to the left child
            elif val > parent.val:  # Traverse to the right subtree
                if parent.right is None:  # Insert if right child is empty
                    parent.right = new_node
                    return
                parent = parent.right  # Move to the right child
            else:  # Duplicate value detected
                print("Duplicate values are not supported")
                return

    def deletion(self, val):
        if self.root is None:
            print("Could not find value")
            return

        # Search for value
        parent = self.root

        if val < parent:
            node = self.root.left
            pointing = "Left"
        else:
            node = self.root.right
            pointing = "Right"

        while node is not None:
            if val > node.val:
                parent = node
                node = node.right
                pointing = "Right"
            elif val < node.val:
                parent = node
                node = node.left
                pointing = "Left"
            else:
                # Deleting a leaf node
                if node.left is None and node.right is None:
                    if pointing == "Left":
                        parent.left = None
                    else:
                        parent.right = None
                    return

                # Deleting a node with only one left child
                if node.right is None:
                    if pointing == "Left":
                        parent.left = node.left
                    else:
                        parent.right = node.left
                else:
                    if pointing == "Left":
                        parent.left = node.right
                    else:
                        parent.right = node.right

                # Deleting a node with only one right child
                if node.left is None:
                    if pointing == "Left":
                        parent.left = node.right
                    else:
                        parent.right = node.right
                else:
                    if pointing == "Left":
                        parent.left = node.left
                    else:
                        parent.right = node.left

                # Deleting a node with 2 children
                if node.left is not None and node.right is not None:
                    if pointing == "Left":
                        node.right.left = node.left
                        parent.left = node.right
                    else:
                        node.left.right = node.right
                        parent.right = node.right
                return
            print("Value not in tree")
            return
            
